clang/lib/Basic/SourceManager.cpp

```diff
@@ -1986,13 +1984,15 @@ bool SourceManager::isInTheSameTranslationUnitImpl(
   if (isLoadedFileID(LOffs.first) != isLoadedFileID(ROffs.first))
     return false;
 
-  // If both are loaded from different AST files.
   if (isLoadedFileID(LOffs.first) && isLoadedFileID(ROffs.first)) {
-    auto FindTableSegment = [this](FileID FID) {
-      return llvm::upper_bound(LoadedSLocEntryTableSegments, -FID.ID - 2);
+    auto FindSLocEntryAlloc = [this](FileID FID) {
+      // FileIDs are negative, we store the beginning of each allocation (the
+      // lowest FileID), later allocations have lower FileIDs.
+      return llvm::upper_bound(LoadedSLocEntryAllocBegin, FID, std::greater{});
```

Comment:
I would have expected lower_bound; is this related to -2 vs -1?

Reply:
No, that's just me not being able to visualize binary search over anything else than positive increasing integers 😅

You're right, this should be `lower_bound`. `upper_bound` would report the incorrect allocation for the first `FileID`. Fixed in the latest commit and added a unit test.

Reply:
But yes, `upper_bound` ended up working because I was incorrectly doing `- 2` instead of `- 1` on the base `FileID`.